src/xdr/impls/generalized_transaction_set.rs

```diff
@@ -0,0 +1,105 @@
+use crate::{
+    compound_types::UnlimitedVarArray,
+    types::{GeneralizedTransactionSet, TransactionPhase, TxSetComponent},
+    Hash, IntoHash, StellarSdkError, TransactionEnvelope, XdrCodec,
+};
+
+#[cfg(feature = "std")]
+impl IntoHash for GeneralizedTransactionSet {
+    fn into_hash(self) -> Result<Hash, StellarSdkError> {
+        use sha2::{Digest, Sha256};
+        use std::convert::TryInto;
+
+        let mut hasher = Sha256::new();
+        hasher.update(self.to_xdr());
+
+        hasher
+            .finalize()
+            .as_slice()
+            .try_into()
+            .map_err(|e: std::array::TryFromSliceError| StellarSdkError::InvalidHashConversion(e.to_string()))
+    }
+}
+
+impl GeneralizedTransactionSet {
+    #[cfg(feature = "std")]
+    pub fn to_base64_encoded_xdr_string(&self) -> Result<String, std::string::FromUtf8Error> {
+        let base_64 = self.clone().to_base64_xdr();
+        String::from_utf8(base_64)
+    }
+
+    pub fn previous_ledger_hash(&self) -> Option<Hash> {
+        let GeneralizedTransactionSet::V1(txset_v1) = self else {
+            return None;
+        };
+        Some(txset_v1.previous_ledger_hash)
+    }
+
+    pub fn txes(&self) -> Option<UnlimitedVarArray<TransactionEnvelope>> {
+        let GeneralizedTransactionSet::V1(txset_v1) = self else {
+            return None;
+        };
+
+        let mut final_txes = UnlimitedVarArray::new_empty();
+
+        for phase in txset_v1.phases.get_vec() {
+            let TransactionPhase::V0(txsets_comp) = phase else {
+                return None;
+            };
+
+            for TxSetComponent::TxsetCompTxsMaybeDiscountedFee(discounted_txs_set) in txsets_comp.get_vec() {
+                if final_txes.len() == 0 {
+                    final_txes = discounted_txs_set.txes.clone();
+                } else {
+                    for tx in discounted_txs_set.txes.get_vec() {
+                        if let Err(_) = final_txes.push(tx.clone()) {
+                            return None
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
Do we need to accumulate them over all phases? 🤔 d
I just want to be sure that we don't add duplicates here. I don't know how exactly these phases work but if it's possible that eg. phase `0` and phase `1` contain the same transactions, we need to change this logic here to remove duplicates.

Did you test this for some 'real' transaction data on testnet @b-yap ?

Reply:
So far when running https://github.com/pendulum-chain/spacewalk/tree/stellar-relay-lib-logs/clients/stellar-relay-lib#example, I did not find duplicates on amongst the phases.

But it's good to add checking, just in case. I'll do that.

Reply:
If the Generalized version is anything like the old transaction set in terms that is one per ledger, then it shouldn't be duplicates, otherwise it would essentially mean double spending no?

Reply:
I was just worried because of the naming with 'phases'. Because without knowing the details, I could have imagined that a transaction goes through more than one phase, which we needed to clarify.
Someone from SDF got back to my question on Slack and it seems like we don't need to worry about it:

> Generalized transaction sets are defined in [CAP-0042](https://github.com/stellar/stellar-protocol/blob/master/core/cap-0042.md): the main idea is to improve the fee mechanism for transactions to get into a block (and control application order of different phases). It sounds like for your purposes, you just need all the transactions that got confirmed in the block, so it seems like it should be fine to simply accumulate all transactions from all phases.


So according to that explanation accumulating the transactions is the right approach and we don't need to worry about duplicates.